nixos/modules/services/networking/ergochat.nix

```diff
@@ -0,0 +1,142 @@
+        '';
+        default = {
+          network = {
+            name = "testnetwork";
+          };
+          server = {
+            name = "example.com";
+            listeners = {
+              ":6667" = {};
+            };
+            casemapping = "permissive";
+            enforce-utf = true;
+            lookup-hostnames = false;
+            ip-cloaking = {
+              enabled = false;
+            };
+            forward-confirm-hostnames = false;
+            check-ident = false;
+            relaymsg = {
+              enabled = false;
+            };
+            max-sendq = "1M";
+            ip-limits = {
+              count = false;
+              throttle = false;
+            };
+          };
+          datastore = {
+            autoupgrade = true;
+            path = "/var/lib/ergo/ircd.db";
+          };
+          accounts = {
+            authentication-enabled = true;
+            registration = {
+              enabled = true;
+              allow-before-connect = true;
+              throttling = {
+                enabled = true;
+                duration = "10m";
+                max-attempts = 30;
+              };
+              bcrypt-cost = 4;
+              email-verification.enabled = false;
+            };
+            multiclient = {
+              enabled = true;
+              allowed-by-default = true;
+              always-on = "opt-out";
+              auto-away = "opt-out";
+            };
+          };
+          channels = {
+            default-modes = "+ntC";
+            registration = {
+              enabled = true;
+            };
+          };
+          limits = {
+            nicklen = 32;
+            identlen = 20;
+            channellen = 64;
+            awaylen = 390;
+            kicklen = 390;
+            topiclen = 390;
+          };
+          history = {
+            enabled = true;
+            channel-length = 2048;
+            client-length = 256;
+            autoresize-window = "3d";
+            autoreplay-on-join = 0;
+            chathistory-maxmessages = 100;
+            znc-maxmessages = 2048;
+            restrictions = {
+              expire-time = "1w";
+              query-cutoff = "none";
+              grace-period = "1h";
+            };
+            retention = {
+              allow-individual-delete = false;
+              enable-account-indexing = false;
+            };
+            tagmsg-storage = {
+              default = false;
+              whitelist = [
+                "+draft/react"
+                "+react"
+              ];
+            };
+          };
+        };
+      };
+
+    };
+  };
+  config = let
+
+    cfg = config.services.ergochat;
+    configFile = pkgs.writeText "ergo.conf" (builtins.toJSON cfg.settings);
+
+  in lib.mkIf cfg.enable ({
+
+    environment.etc."ergo.yaml".source = configFile;
+    services.ergochat.settings =
+      lib.mapAttrsRecursive (_: lib.mkDefault) options.services.ergochat.settings.default;
```

Comment:
This way of settings defaults might need an explanatory comment as it doesn't seem to be a widely used construct. (If used at all in nixpkgs--I couldn't find anything similar.)

Reply:
not sure it's a great user experience since this config is pretty large and we have no `mkUnset` in the module system. maybe split `settings` into two attributes (perhaps `baseSettings` and `settings`) and merge during config file generation, or use only the non-defaulted values if the user requests it.

Reply:
IMHO this is the expected user experience, this is a good compromise between the minimal config and the default config shipped by ergo. if you don't want some features you can just disable them. for example via `settings.history.enabled = false` and all history related features don't have any effect anymore. if you want to have the hole history section gone you can say `settings.history = lib.mkForce {};`

Reply:
with that method you could also replace the entire config. yeah, makes sense.